alt_e2eshark/onnx_tests/models/model.py

```diff
@@ -0,0 +1,209 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+import os
+import numpy
+import onnx
+import torch
+from onnxruntime.tools.onnx_model_utils import make_dim_param_fixed, fix_output_shapes
+from e2e_testing import azutils
+from e2e_testing.framework import OnnxModelInfo
+from e2e_testing.registry import register_test
+from e2e_testing.storage import TestTensors
+from .protected_list import protected_models, models_with_postprocessing
+
+class AzureDownloadableModel(OnnxModelInfo):
+    def __init__(self, name: str, onnx_model_path: str, cache_dir: str):
+        opset_version = 21
+        super().__init__(name, onnx_model_path, cache_dir, opset_version)
+
+    def construct_model(self):
+        # try to find a .onnx file in the test-run dir
+        # if that fails, check for zip file in cache
+        # if that fails, try to download and setup from azure, then search again for a .onnx file
+
+        # TODO: make the zip file structure more uniform so we don't need to search for extracted files
+        model_dir = self.model.rstrip("model.nx")
+
+        def find_models(model_dir):
+            # search for a .onnx file in the ./test-run/testname/ dir
+            found_models = []
+            for root, dirs, files in os.walk(model_dir):
+                for name in files:
+                    if name[-5:] == ".onnx":  
+                        found_models.append(os.path.abspath(os.path.join(root, name)))
+            return found_models
+
+        found_models = find_models(model_dir) 
+
+        if len(found_models) == 0:
+            azutils.pre_test_onnx_model_azure_download(
+                self.name, self.cache_dir, self.model
+            )
+            found_models = find_models(model_dir)
+        if len(found_models) == 1:
+            self.model = found_models[0]
+            return
+        if len(found_models) > 1:
+            print(f'Found multiple model.onnx files: {found_models}')
+            print(f'Picking the first model found to use: {found_models[0]}')
+            self.model = found_models[0]
+            return
+        raise OSError(f"No onnx model could be found, downloaded, or extracted to {model_dir}")
+
+    def apply_postprocessing(self, output: TestTensors):
+        if self.name not in models_with_postprocessing:
+            return output
+        processed_outputs = []
+        for d in output.to_torch().data:
+            processed_outputs.append(
+                torch.sort(torch.topk(torch.nn.functional.softmax(d, 1), 2)[1])[0]
+            )
+        return TestTensors(processed_outputs)
+
+for t in protected_models:
+    register_test(AzureDownloadableModel, t)
+
+
+register_test(AzureDownloadableModel, "RAFT_vaiq_int8")
+register_test(AzureDownloadableModel, "pytorch-3dunet_vaiq_int8")
+register_test(AzureDownloadableModel, "FCN_vaiq_int8")
+register_test(AzureDownloadableModel, "u-net_brain_mri_vaiq_int8")
```

Comment:
Why multiple different models in a single test file? How about one model per file, with multiple test _cases_ in each file (e.g. different inputs or functions to the same program)?

Reply:
What would be in the file? 
```python
import stuff
register_test(AzureDownloadableModel, "modelname")
# end of file
```

I think it would be fine to split it up thematically like `azure_model.py`, `hf_vision_model.py` etc. But having potentially thousands of separate files just to add individual tests seemed like something I wanted to avoid. Is there a benefit to keeping the files separate for individual tests? E.g., does pytest require this structure?

The test-run directory that get generated when running tests does break up results/artifacts/logs into separate directories, if that is helpful context.

Reply:
If there are 1000s of test models each with 50 lines of code, I'd much rather have 1000s of test files than have a single 50000 line test file :P

Test runners/frameworks like pytest group by package, then file, then test case. They let you run all tests in a package, all tests in a file, a single test case, etc.

Reply:
I don't think there is anything preventing this structure from having more files if they are getting too cluttered. The part that takes up the most space is definitely defining a new OnnxModelInfo child class, so it would make sense to separate those out into different files.

My feeling was that, since most models are available on azure for downloading directly, adding a new test would really be just specifying the name. I'd rather have:
```python
for t in some_list_of_testnames:
   register_test(AzureDownloadableModel, t)
```
with the list of testnames gotten from like a text file or json, than having thousands of files, or thousands of lines of any kind of code.

In any case, I can demo this change on the operator tests, since that model.py is already getting cluttered.

Reply:
I see your point about 
```python
for t in some_list_of_testnames:
   register_test(AzureDownloadableModel, t)
```
However, code like `Opt125MGPTQModelInfo` should go in its own file.

It's also nice being able to look at the directory tree and grep / file name search to find tests, but a monolithic `model.py` (should that be `models.py`?) needs its contents searched for a similar query. Collaboration like multiple people adding test cases is also easier with split files - lower risk of merge conflicts.